carma_wm_ctrl/src/WMBroadcaster.cpp

```diff
@@ -197,104 +296,330 @@ std::shared_ptr<Geofence> WMBroadcaster::geofenceFromMsg(const cav_msgs::Traffic
+      return {};
+    }
+    else
+    {
+      return_list = createWorkzoneGeofence(work_zone_geofence_cache_);
+      return return_list;
+    }
+  }
+  else if (msg_detail.choice == cav_msgs::TrafficControlDetail::CLOSED_CHOICE && msg_detail.closed==cav_msgs::TrafficControlDetail::CLOSED) // if stand-alone closed signal apart from Workzone
+  {
+    addRegionAccessRule(gf_ptr,msg_v01,affected_llts);
+  }
+ 
+  return_list.push_back(gf_ptr);
+  return return_list;
+}
+
+std::vector<std::shared_ptr<Geofence>> WMBroadcaster::createWorkzoneGeofence(std::unordered_map<uint8_t, std::shared_ptr<Geofence>> work_zone_geofence_cache)
+{
+  std::shared_ptr<std::vector<lanelet::Lanelet>> parallel_llts = std::make_shared<std::vector<lanelet::Lanelet>>(std::vector<lanelet::Lanelet>());
+  std::shared_ptr<std::vector<lanelet::Lanelet>> middle_opposite_lanelets = std::make_shared<std::vector<lanelet::Lanelet>>(std::vector<lanelet::Lanelet>());
   
-  ros::Time end_time = msg_schedule.end;
-  if (!msg_schedule.end_exists) {
-    ROS_DEBUG_STREAM("No end time for geofence, using ros::TIME_MAX");
-    end_time = ros::TIME_MAX; // If there is no end time use the max time
+  // Split existing lanelets and filter into parallel_llts and middle_opposite_lanelets
+  preprocessWorkzoneGeometry(work_zone_geofence_cache, parallel_llts, middle_opposite_lanelets);
+
+  // Create geofence and rest of the required lanelets along with traffic light and stop rule for completing workzone area
+  auto gf_ptr = createWorkzoneGeometry(work_zone_geofence_cache, parallel_llts->front(), parallel_llts->back(), middle_opposite_lanelets );
+
+  // copy static info from the existing workzone
+  gf_ptr->id_ = work_zone_geofence_cache[cav_msgs::TrafficControlDetail::TAPERRIGHT]->id_; //using taperright's id as the whole geofence's id
+  // schedule
+  gf_ptr->schedules = work_zone_geofence_cache[cav_msgs::TrafficControlDetail::TAPERRIGHT]->schedules; //using taperright's schedule as the whole geofence's schedule
+  return {gf_ptr};
+}
+
+std::shared_ptr<Geofence> WMBroadcaster::createWorkzoneGeometry(std::unordered_map<uint8_t, std::shared_ptr<Geofence>> work_zone_geofence_cache,  lanelet::Lanelet parallel_llt_front,  lanelet::Lanelet parallel_llt_back, 
+                                                    std::shared_ptr<std::vector<lanelet::Lanelet>> middle_opposite_lanelets)
+{
+  auto gf_ptr = std::make_shared<Geofence>(Geofence());
+
+  //create trapezoid llt on front side TODO: more sophisticated diagonal llt?
+  lanelet::LineString3d front_left_ls(lanelet::utils::getId(), {parallel_llt_front.leftBound3d().back(), middle_opposite_lanelets->back().rightBound3d().back()});
+  lanelet::LineString3d front_right_ls(lanelet::utils::getId(), {parallel_llt_front.rightBound3d().back(), middle_opposite_lanelets->back().leftBound3d().back()});
+  lanelet::Lanelet front_llt_diag(lanelet::utils::getId(), front_left_ls, front_right_ls);
```

Comment:
I think we need some linear interpolation on these diagonals such that the points end up spaced at 1/4m increments (configurable) like our maps. That will let it play nice with inlane-cruising

Reply:
added interpolating lanelets that have configurable linestring point increments